packages/@aws-cdk/aws-codebuild/lib/project.ts

```diff
@@ -764,59 +764,97 @@ export class Project extends ProjectBase {
             // KMS on the other account as well to be able to get the secret
-            if (parsedArn && parsedArn.account && Token.compareStrings(parsedArn.account, stack.account) === TokenComparison.DIFFERENT) {
+            if (parsedArn.account && Token.compareStrings(parsedArn.account, stack.account) === TokenComparison.DIFFERENT) {
               kmsIamResources.add(stack.formatArn({
                 service: 'kms',
                 resource: 'key',
                 // We do not know the ID of the key, but since this is a cross-account access,
                 // the key policies have to allow this access, so a wildcard is safe here
                 resourceName: '*',
                 sep: '/',
-                // if we were given an ARN, we need to use the provided partition/account/region
                 partition: parsedArn.partition,
                 account: parsedArn.account,
                 region: parsedArn.region,
               }));
             }
+          } else if (Token.isUnresolved(envVariableValue)) {
+            // the value of the property can be a complex string, separated by ':';
+            // see https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html#build-spec.env.secrets-manager
+            let secretArn = envVariableValue.split(':')[0];
+
+            // parse the Token, and see if it represents a single resource
+            // (we will assume it's a Secret from SecretsManager)
+            const fragments = Tokenization.reverseString(envVariableValue);
+            if (fragments.tokens.length === 1) {
+              const iresolvable = fragments.tokens[0];
+              if (Reference.isReference(iresolvable)) {
+                // check the Stack the resource owning the reference belongs to
+                const resourceStack = Stack.of(iresolvable.target);
+                if (Token.compareStrings(stack.account, resourceStack.account) === TokenComparison.DIFFERENT) {
+                  // since this is a cross-account access,
+                  // add the appropriate KMS permissions
+                  kmsIamResources.add(stack.formatArn({
+                    service: 'kms',
+                    resource: 'key',
+                    // We do not know the ID of the key, but since this is a cross-account access,
+                    // the key policies have to allow this access, so a wildcard is safe here
+                    resourceName: '*',
+                    sep: '/',
+                    partition: resourceStack.partition,
+                    account: resourceStack.account,
+                    region: resourceStack.region,
+                  }));
+
+                  // Work around a bug in SecretsManager -
+                  // when the access is cross-environment,
+                  // Secret.secretArn returns a partial ARN!
+                  // So add a "*" at the end, so that the permissions work
+                  secretArn = `${secretArn}*`;
```

Comment:
Wondering whether this should be `-??????` maybe?

```suggestion
                  secretArn = `${secretArn}-??????`;
```

Reply:
Or maybe no, because it wouldn't work for imported Secrets by complete ARN...? But maybe that doesn't use a Token...? Another test is needed perhaps...? 🙂

Reply:
Hm wouldn't we expect that if someone tries to import a secret by complete ARN that he would just pass it as the value directly? Then there would be no need for adding the extra test and we can just change this to `-??????`

While thinking about this I came up with another test case. What would happen if the `secretArn` is only available as a `Token`. Would we be able to figure out if it comes from another account?
@skinny85 I have a feeling this is the thing we are missing from the original idea. I hope I am wrong here and you can correct me on this ^^

Reply:
> Hm wouldn't we expect that if someone tries to import a secret by complete ARN that he would just pass it as the value directly? Then there would be no need for adding the extra test and we can just change this to -??????

There might be other reasons they are using a Secret (for example, they also reference it in the Stack that imports it). So I think a new test is justified here.

> While thinking about this I came up with another test case. What would happen if the secretArn is only available as a Token. Would we be able to figure out if it comes from another account?
@skinny85 I have a feeling this is the thing we are missing from the original idea. I hope I am wrong here and you can correct me on this ^^

Not sure I understand. Can you show me a unit test that is the problem here?

Reply:
Alright so I added the test you asked for `can be provided as a SecretArn of a Secret imported by complete ARN from a different account`.

And also the test that should shows what I mean what the problem is `can be provided as a SecretArn of a Secret imported by complete ARN provided as a Token from a different account`.

Reply:
> And also the test that should shows what I mean what the problem is `can be provided as a SecretArn of a Secret imported by complete ARN provided as a Token from a different account`.

Do you mean this?

```ts

      'can be provided as a SecretArn of a Secret imported by complete ARN provided as a Token from a different account'(test: Test) {
        // GIVEN
        const app = new cdk.App();
        const secretArnToken = cdk.Token.asString({ resolve: () => 'arn:aws:secretsmanager:us-west-2:901234567890:secret:mysecret' });
        const stack = new cdk.Stack(app, 'ProjectStack', {
          env: { account: '123456789012' },
        });

        // WHEN
        const secret = secretsmanager.Secret.fromSecretCompleteArn(stack, 'Secret', secretArnToken);
        new codebuild.PipelineProject(stack, 'Project', {
          environmentVariables: {
            'ENV_VAR1': {
              type: codebuild.BuildEnvironmentVariableType.SECRETS_MANAGER,
              value: `${secret.secretArn}:json-key`,
            },
          },
        });
```

I don't understand the point of this test, honestly... what use-case is this supposed to represent?

Reply:
Yes I mean that test exactly. Let me try to explain what I am trying to solve.
Let's say in our Stack we use a `CustomResource`, which creates a `Secret` in another account and provides the secretArn through the `GetAtt` intrinsic function. The code would look something like this I reckon:
```ts
const customResource = new CustomResource(this, 'CustomResource');
const secretArn = customResource.secretArn //This is a token and resolves to { "Fn::GetAtt": [ "CustomResource", "SecretArn" ] }
```

In this event our current setup won't work.
```ts
const fragments = Tokenization.reverseString(envVariableValue);
if (fragments.tokens.length === 1) {
  const resolvable = fragments.tokens[0];
    if (Reference.isReference(resolvable)) {
    // check the Stack the resource owning the reference belongs to
    const resourceStack = Stack.of(resolvable.target);
    if (Token.compareStrings(stack.account, resourceStack.account) === TokenComparison.DIFFERENT) {
      ...
    }
  }
}
```
`resourceStack.account` will be the same account as `stack.account`. I think we would have to check the secretArn provided to us to identify that the secret comes from a different account. This could look something like this:
```ts
const fragments = Tokenization.reverseString(envVariableValue);
if (fragments.tokens.length === 1) {
  const parsedArn = stack.parseArn(secretArn);
  if (Token.compareStrings(stack.account, parsedArn.account) === TokenComparison.DIFFERENT) {
    ...
  }
}
```
Although when we do this the account parsed from the secretArn will probably look something like this:
```ts
{
  "Fn::Select": [4, {
    "Fn::Split": [":",
      {
        "Fn::GetAtt": ["CustomResource", "SecretArn"],
      }],
  }],
},
```
I believe this would result in the token comparison to always pass right?

---
@skinny85 I hope this makes my thought process clear. Do you think this is something we should try to solve in cdk?

Reply:
I don't think this is solvable in the CDK at all - we will never know that `{  "Fn::GetAtt": ["CustomResource", "SecretArn"] }` actually points to something in a different account than the account the Project is in (there is no account to compare the Project's account to).

Does this mean this PR does not solve your problem?

Reply:
Well at least it does not solve the case that we provide the KMS policy every time it's needed. That's what I was aiming for.

I think we still did a lot of improvements in this PR.

Maybe it's worth adding some documentation about cases where one would have to add KMS policies on their own?

Reply:
Feel free to take a stab at it in this PR 🙂.

Reply:
I've been thinking about this some more.

> Let's say in our Stack we use a CustomResource, which creates a Secret in another account and provides the secretArn through the GetAtt intrinsic function.

Are you saying that the Custom Resource lives in account A, but creates a Secret in account B?

Reply:
> Are you saying that the Custom Resource lives in account A, but creates a Secret in account B?

Yeah exactly. That's one of the use cases we have. What were your thoughts?

Reply:
My thought is that the reason you're running into this problem that you're breaking the CloudFormation contract.

If a CloudFormation Stack is in account A and region R, the assumption is that all resources inside of it are also in account A and region R.

But you're breaking that assumption, and suddenly have a resource in a Stack in account A that actually lives in account B. I think that's the source of your problems here.

I would consider a design where the Custom Resource you have creating the Secret has to live in the same account as the account it creates the Secret in (region as well).

If you did it this way, I'm pretty sure the code of this PR would work for your use-case.

Thoughts?

Reply:
Yeah I think your reasoning is valid. I will discuss this and remove the showcase test for now so we can merge this PR. 

Thanks for all your input and help :)